src/prepared_statement.c

```diff
@@ -0,0 +1,466 @@
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashtab-impl.h>
+
+static PgPreparedStatement* prepared_statements;
+queryid_t next_unique_query_id;
+
+/*
+ * Like uthash HASH_DELETE, but doesn't remove item from hash table just
+ * unlink element from the doubly-linked-list.
+ */
+#define HASH_UNLINK(hh, head, delptr)									\
+	do {																\
+		struct UT_hash_handle *_hd_hh_del = &(delptr)->hh;				\
+		if (_hd_hh_del->prev != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev)->next = _hd_hh_del->next; \
+		} else {														\
+			DECLTYPE_ASSIGN(head, _hd_hh_del->next);					\
+		}																\
+		if (_hd_hh_del->next != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->next)->prev = _hd_hh_del->prev; \
+		} else {														\
+			_hd_hh_del->tbl->tail = HH_FROM_ELMT(_hd_hh_del->tbl, _hd_hh_del->prev); \
+		}																\
+	} while (0)
+
+static PgPreparedStatement*
+create_prepared_statement(PgParsePacket *pkt)
+{
+	size_t query_len = strlen(pkt->query) + 1;
+	size_t num_parameters = pkt->num_parameters * 4;
+	PgPreparedStatement *ps = (PgPreparedStatement *)malloc(sizeof(PgPreparedStatement) + query_len + num_parameters);
+	ps->query_len = query_len;
+	ps->param_len = num_parameters;
+	ps->query_id = next_unique_query_id;
+	ps->use_count = 0;
+	memcpy(ps->query_with_params, pkt->query, query_len);
+	memcpy(&ps->query_with_params[query_len], pkt->parameter_types_bytes, num_parameters);
+	return ps;
+}
+
+static PgClientPreparedStatement*
+create_client_prepared_statement(char const* name, PgPreparedStatement* ps)
+{
+	size_t name_len = strlen(name) + 1;
+	PgClientPreparedStatement *client_ps = (PgClientPreparedStatement*)malloc(sizeof(PgClientPreparedStatement) + name_len);
+	memcpy(client_ps->stmt_name, name, name_len);
+	client_ps->ps = ps;
+	ps->use_count += 1;
+	return client_ps;
+}
+
+static PgServerPreparedStatement*
+create_server_prepared_statement(PgPreparedStatement* ps)
+{
+	PgServerPreparedStatement *server_ps = (PgServerPreparedStatement*)malloc(sizeof(PgServerPreparedStatement));
```

Comment:
Since this is fixed size lets create a slab for it.

Reply:
Done

Reply:
CI is reporting some valgrind errors because of this change I think.

```
==12835== VALGRIND-ERROR-BEGIN
==12835== Invalid free() / delete / delete[] / realloc()
==12835==    at 0x483CA3F: free (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==12835==    by 0x13DB9F: ps_server_free (prepared_statement.c:473)
==12835==    by 0x12B144: change_server_state (objects.c:271)
==12835==    by 0x12D4DF: disconnect_server (objects.c:1074)
==12835==    by 0x123482: close_server_list (janitor.c:39)
==12835==    by 0x1254F2: kill_pool (janitor.c:754)
==12835==    by 0x1256FC: kill_database (janitor.c:795)
==12835==    by 0x1308A3: objects_cleanup (objects.c:2191)
==12835==    by 0x129E50: cleanup (main.c:856)
==12835==    by 0x4D028A6: __run_exit_handlers (exit.c:108)
==12835==    by 0x4D02A5F: exit (exit.c:139)
==12835==    by 0x128B36: handle_sigterm (main.c:475)
==12835==  Address 0x51a20d0 is 16 bytes inside a block of size 16,360 alloc'd
==12835==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==12835==    by 0x15E041: libc_alloc (cxalloc.c:148)
==12835==    by 0x15DA88: cx_alloc (cxalloc.c:34)
==12835==    by 0x15DBD6: cx_alloc0 (cxalloc.c:69)
==12835==    by 0x161C96: grow (slab.c:149)
==12835==    by 0x161D7B: slab_alloc (slab.c:173)
==12835==    by 0x136E5F: create_server_prepared_statement (prepared_statement.c:74)
==12835==    by 0x13A441: handle_parse_command (prepared_statement.c:208)
==12835==    by 0x11F29B: handle_client_work (client.c:1069)
==12835==    by 0x11F741: client_proto (client.c:1157)
==12835==    by 0x13F6B9: sbuf_call_proto (sbuf.c:408)
==12835==    by 0x1401DD: sbuf_process_pending (sbuf.c:608)
==12835== 
==12835== VALGRIND-ERROR-END
```

Reply:
Sorry, fixed